students/Gregg/lesson3/mailroom.py

```diff
@@ -1,148 +1,149 @@
 #!/usr/bin/env python3
 """Use current knowledge to set up an interactive program"""
 
-donations = {
-    "William Gates, III": [3, 5, 7],
-    "Mark Zuckerberg": [4.50, 8, 2],
-    "Jeff Bezos": [7.77],
-    "Paul Allen": [3.6, 4.5],
-    "bob": [.01]
-}
-return_to_primary_prompt = [False]
-
-# This doesn't seem like the cleanest way to do this, but I coultn't
-# come up with anything better without turning it into a full blow stat machine
-# that has a centralized handler thingy for every possible transition
-# That can then return to the start state if necessary
-# It seems like it works ok for this limited example
+#Todo - restructure so there is only one message handler, with dict mapping prompt, input options, functions and next states
+#Todo - input a dictionary to the letter writing function
+# this is a pain because I need to make the dictionary just to take it apart...it
+#Todo add option to save all leters
+#need additional save function
+
+
+donors = [
+    "William Gates, III",
+    "Mark Zuckerberg",
+    "Jeff Bezos",
+    "Paul Allen",
+    "bob"
+]
+
+donations = [
+    [3, 5, 7],
+    [4.50, 8, 2],
+    [7.77],
+    [3.6, 4.5],
+    [40]
+]
```

Comment:
the above data structure over-complicates your code quite a bit

this is the common data structure:
```
donor_db = [("William Gates, III", [653772.32, 12.17]),
            ("Jeff Bezos", [877.33]),
            ("Paul Allen", [663.23, 43.87, 1.32]),
            ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
            ]
```
as mentioned in here:
https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom_tutorial.html#choosing-a-data-structure


this is how simple your report generation can look:
```
def print_donor_report():
    """
    Generate the report of the donors and amounts donated.
    """
    # First, reduce the raw data into a summary list view
    report_rows = []
    for name, gifts in donor_db:
        total_gifts = sum(gifts)
        num_gifts = len(gifts)
        avg_gift = total_gifts / num_gifts
        report_rows.append((name, total_gifts, num_gifts, avg_gift))

    # sort the report data
    report_rows.sort(key=sort_key)

    # print it out in with a nice format.
    print("{:25s} | {:11s} | {:9s} | {:12s}".format(
          "Donor Name", "Total Given", "Num Gifts", "Average Gift"))
    print("-" * 66)
    for row in report_rows:
        print("{:25s}   {:11.2f}   {:9d}   {:12.2f}".format(*row))

```